File: src/main/java/com/github/sblundy/elvish/psi/ElvishVariableRefBase.java
package com.github.sblundy.elvish.psi;

import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import com.intellij.psi.*;
import org.jetbrains.annotations.NotNull;

public abstract class ElvishVariableRefBase extends ASTWrapperPsiElement implements ElvishVariableRef {
    public ElvishVariableRefBase(@NotNull ASTNode node) {
        super(node);
    }

    @Override
    public PsiReference getReference() {
        return new ElvishVariableReference(this, getVariableName().getTextRangeInParent());
    }
}

File: src/main/kotlin/com/github/sblundy/elvish/psi/ElvishVariableRefBase.kt
package com.github.sblundy.elvish.psi

import com.intellij.lang.ASTNode
import com.intellij.psi.PsiReference

abstract class ElvishVariableRefBase(node: ASTNode) : ASTWrapperElvishPsiElement(node), ElvishVariableRef {
    override fun getReference(): PsiReference? = ElvishVariableReference(this, variableName.textRangeInParent)
}